# Flip Robo Technologies Assignment - Web Scraping Beautiful soup
## Solved by Mihir Bhoite Batch - DS2406

### In all the following questions, you have to use BeautifulSoup to scrape different websites and collect data as per the requirement of the question.


#### Question 1) Write a python program to display IMDB’s Top rated 100 Indian movies’ data https://www.imdb.com/list/ls056092300/ (i.e. name, rating, year ofrelease) and make data frame.

In [5]:
!pip install bs4

  Obtaining dependency information for bs4 from https://files.pythonhosted.org/packages/51/bb/bf7aab772a159614954d84aa832c129624ba6c32faa559dfb200a534e50b/bs4-0.0.2-py2.py3-none-any.whl.metadata


In [6]:
!pip install requests

In [9]:
from bs4 import BeautifulSoup 
import requests

In [10]:
page = requests.get('https://www.imdb.com/list/ls056092300/')

In [11]:
page

<Response [403]>

In [12]:
soup = BeautifulSoup(page.content)
soup


<html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
</body>
</html>

##### as here above we checked we are getting the response in the range of 400-500. That means web scraping is not possible in the particular page (Imdb) as it showing reponse 403, means it is forbidden to do so. 

#### Question 2) Write a python program to scrape details of all the posts from https://www.patreon.com/coreyms .Scrape the heading, date, content and the likes for the video from the link for the youtube video from the post.


In [13]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

# URL 
url = 'https://www.patreon.com/coreyms'

# Send GET request to the URL
response = requests.get(url)
soup = BeautifulSoup(response.content, 'html.parser')

posts = []

post_data = soup.find_all('div', class_='post-class')  

for post in post_data:
    heading = post.find('h2', class_='post-title-class').text.strip() 
    date = post.find('time', class_='post-date-class').text.strip()    
    content = post.find('div', class_='post-content-class').text.strip()  
    likes = post.find('span', class_='likes-class').text.strip()        
    video_link = post.find('a', class_='youtube-link-class')['href']   

    posts.append({
        'Heading': heading,
        'Date': date,
        'Content': content,
        'Likes': likes,
        'YouTube Link': video_link
    })

df = pd.DataFrame(posts)

# Display the DataFrame
print(df)



Empty DataFrame
Columns: []
Index: []


In [14]:
page = requests.get('https://www.patreon.com/coreyms')
page

<Response [200]>

In [15]:
soup = BeautifulSoup(page.content)
soup

<!DOCTYPE html>
<html><head><meta charset="utf-8"/><script type="text/javascript">
            window.patreonConfig = {"apiServer":"www.patreon.com/api","nionDisableDefaultIncludes":true,"formattingLocale":"en-US","eventLoggingPlatform":"Web","isWebView":false,"pathname":"/\u003cstring:campaign_identifier\u003e","is_impersonating":false,"csrfSignature":"tdP2qKq6EBqjxsUAq0C1OI4vUmZBD4mECDSjhhPrU7Q"};
          </script><meta content="width=device-width, initial-scale=1.0, viewport-fit=cover" name="viewport"/><meta content="artists, creators, patron, sponsor, music, videos" name="keywords"/><meta content="201564" property="fb:admins"/><meta content="130127590512253" property="fb:app_id"/><meta content="IE=edge" http-equiv="X-UA-Compatible"/><meta content="@Patreon" property="twitter:site"/><meta content="Patreon" property="og:site-name"/><meta content="Patreon" prefix="og: http://ogp.me/ns#" property="og:site_name"/><meta content="Patreon" name="apple-mobile-web-app-title"/><meta content

In [17]:
!pip install selenium

  Obtaining dependency information for selenium from https://files.pythonhosted.org/packages/c9/33/b9da8be5b122b8c3c82c35f515ba0a84a9af3ba9629ae9fd5bbba820d592/selenium-4.23.1-py3-none-any.whl.metadata
  Obtaining dependency information for trio~=0.17 from https://files.pythonhosted.org/packages/1c/70/efa56ce2271c44a7f4f43533a0477e6854a0948e9f7b76491de1fd3be7c9/trio-0.26.2-py3-none-any.whl.metadata
  Obtaining dependency information for trio-websocket~=0.9 from https://files.pythonhosted.org/packages/48/be/a9ae5f50cad5b6f85bd2574c2c923730098530096e170c1ce7452394d7aa/trio_websocket-0.11.1-py3-none-any.whl.metadata
  Obtaining dependency information for typing_extensions~=4.9 from https://files.pythonhosted.org/packages/26/9f/ad63fc0248c5379346306f8668cda6e2e2e9c95e01216d2b8ffd9ff037d0/typing_extensions-4.12.2-py3-none-any.whl.metadata
  Obtaining dependency information for websocket-client~=1.8 from https://files.pythonhosted.org/packages/5a/84/44687a29792a70e111c5c477230a72c4b957d88d16

In [25]:
import requests
from bs4 import BeautifulSoup
import pandas as pd


url = 'https://www.patreon.com/coreyms'
response = requests.get(url)
soup = BeautifulSoup(response.content, 'html.parser')

posts = soup.find_all('div', class_='post')

for post in posts:
  heading = post.find('h3').text
  date = post.find('time')['datetime']
  content = post.find('div', class_='post__content').text
  youtube_link = post.find('a', class_='post__youtube-link')['href']
  likes = post.find('span', class_='post__likes').text

  print(f'Heading: {heading}')
  print(f'Date: {date}')
  print(f'Content: {content}')
  print(f'YouTube Link: {youtube_link}')
  print(f'Likes: {likes}')
  print('\n')


#### Question-3 Write a python program to scrape house details from mentioned URL. It should include house title, location,area, EMI and price from https://www.nobroker.in/ .Enter three localities which are Indira Nagar, Jayanagar, Rajaji Nagar.

In [29]:
page = requests.get('https://www.nobroker.in/')
page

<Response [200]>

In [31]:
from bs4 import BeautifulSoup
import requests
import pandas as pd


api_url = "https://www.nobroker.in/api/v1/multi/property/sale/filter"

query = {
    "pageNo": 0,
    "searchParam": "W3sibGF0IjoxMy4wMDExNzc0LCJsb24iOjgwLjI1NjQ5NTcsInBsYWNlSWQiOiJDaElKZ1JiRUZlMW5Vam9SZzU0a2VwYk9hV1UiLCJwbGFjZU5hbWUiOiJBZHlhciJ9XQ==",
    "radius": "2.0",
    "city": ["chennai", "chennai"],
    "locality": "Adyar",
}


all_data, page = [], 0
while True:
    print("Getting page. {}".format(page))
    query["pageNo"] = page
    data = requests.get(api_url, params=query).json()

    if not data["data"]:
        break

    # uncomment this to print all data:
    # print(json.dumps(data, indent=4))

    for p in data["data"]:
        all_data.append(
            {
                "house_title": p["propertyTitle"],
                "location": p["locality"],
                "emi": p["defaultEmi"],
                "price": p["price"],
            }
        )

    page += 1

df = pd.DataFrame(all_data)
print(df)


Getting page. 0
Getting page. 1
Getting page. 2
Getting page. 3
Getting page. 4
Getting page. 5
Getting page. 6
                                           house_title  \
0       3 BHK Flat In Tvh Quadrant, For Sale  In Adyar   
1       2 BHK Flat In Avenue Vistaa For Sale  In Adyar   
2    4 BHK Flat In Jyeshta Apartment  For Sale  In ...   
3                2 BHK Flat For Sale  In Thiruvanmiyur   
4    2 BHK Flat In Ananti Constructions For Sale  I...   
..                                                 ...   
139                          2 BHK For Sale  In Adayar   
140  3 BHK Flat In Shastri Nagar, Adyar For Sale  I...   
141          2 BHK House For Sale  In Padmanabha Nagar   
142  2 BHK Flat In Logu For Sale  In Adyar Shastri ...   
143                    4+ BHK House For Sale  In Adyar   

                                    location              emi     price  
0                                      Adyar  3.61 Lacs/Month  63000000  
1                                      Adya

#### Question-4) Write a python program to scrape first 10 product details which include product name , price , Image URL from
https://www.bewakoof.com/bestseller?sort=popular .

In [34]:
page = requests.get('https://www.bewakoof.com/bestseller?sort=popular')
page

<Response [200]>

In [38]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

# URL 
url = 'https://www.bewakoof.com/bestseller?sort=popular'

response = requests.get(url)

# Parse 
soup = BeautifulSoup(response.content, 'html.parser')

product_containers = soup.find_all('div', class_='productCardBox')[:10]

# Initialize a list 
products = []

for product in product_containers:
    # Extract the product name
    product_name = product.find('h3')
    product_name = product_name.text if product_name else 'N/A'
    
    # Extract the price
    price = product.find('span', class_='discountedPriceText')
    price = price.text if price else 'N/A'
    
    # Extract the image URL
    image = product.find('img')
    image_url = image['src'] if image else 'N/A'

    products.append({
        'Product Name': product_name,
        'Price': price,
        'Image URL': image_url
    })


df = pd.DataFrame(products)

# Print the DataFrame
print(df)


                     Product Name Price  \
0                       Bewakoof®   N/A   
1                       Bewakoof®   N/A   
2  bewakoof x house of the dragon   N/A   
3          bewakoof x tom & jerry   N/A   
4                 bewakoof x nasa   N/A   
5                       Bewakoof®   N/A   
6             bewakoof x garfield   N/A   
7               bewakoof x disney   N/A   
8                       Bewakoof®   N/A   
9                       Bewakoof®   N/A   

                                           Image URL  
0  https://images.bewakoof.com/t640/women-s-pink-...  
1  https://images.bewakoof.com/t640/women-s-red-i...  
2  https://images.bewakoof.com/t640/men-s-black-h...  
3  https://images.bewakoof.com/t640/women-s-black...  
4  https://images.bewakoof.com/t640/men-s-red-nas...  
5  https://images.bewakoof.com/t640/men-s-black-e...  
6  https://images.bewakoof.com/t640/women-s-pink-...  
7  https://images.bewakoof.com/t640/women-s-pink-...  
8  https://images.bewakoof.com/

#### Question-5) Please visit https://www.cnbc.com/world/?region=world and scrap -
 a) headings
b) date
c) News link

In [39]:
page = requests.get('https://www.cnbc.com/world/?region=world')
page

<Response [200]>

In [50]:
import requests
from bs4 import BeautifulSoup
import pandas as pd


url = "https://www.cnbc.com/world/?region=world"
response = requests.get(url)

soup = BeautifulSoup(response.content, "html.parser")

articles = soup.find_all("div", class_="Card-titleContainer")


headlines = []
times = []
links = []

for article in articles:
    # Extract the headline
    headline_element = article.find("a")
    headline = headline_element.text.strip() if headline_element else None
    headlines.append(headline)
    
    # Extract the time
    time_element = article.find("time")
    time = time_element.text.strip() if time_element else 'No time available'
    times.append(time)
    
    # Extract the news link
    link = headline_element["href"] if headline_element else None
    links.append(link)


data = {
    "Headline": headlines,
    "Time": times,
    "News Link": links
}
df = pd.DataFrame(data)

# Print the df
print(df)


                                             Headline               Time  \
0   S&P 500 rises after Fed minutes signal likelih...  No time available   
1   Fed minutes point to ‘likely’ rate cut coming ...  No time available   
2   U.S. job growth revised down by the most since...  No time available   
3   Treasury yields fall after Fed minutes point t...  No time available   
4   'Get Britain building again': New UK finance c...  No time available   
5   Britain's Labour pulled off a thumping electio...  No time available   
6   UK's Labour Party secures landslide victory in...  No time available   
7   Labour does not have much headroom in terms of...  No time available   
8   Outgoing UK PM Rishi Sunak to step down as Con...  No time available   
9   This startup will sell methane-eating microbes...  No time available   
10  Air New Zealand becomes first major airline to...  No time available   
11  World shatters heat record for the second time...  No time available   
12  World lo

#### Question-6) Please visit https://www.keaipublishing.com/en/journals/artificial-intelligence-in-agriculture/most-downloadedarticles/ and scrap-
a) Paper title
 b) date
c) Author

In [51]:
page = requests.get('https://www.keaipublishing.com/en/journals/artificial-intelligence-in-agriculture/most-downloadedarticles/')
page

<Response [404]>

##### as here above we checked we are getting the response in the range of 400-500. That means web scraping is not possible in the particular page (keaipublishing) as it showing reponse 404, means it is forbidden to do so. 